pkg/compressor/compressor.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compressor
+
+import (
+	"compress/gzip"
+	"compress/lzw"
+	"compress/zlib"
+	"fmt"
+	"io"
+
+	"github.com/sirupsen/logrus"
+)
+
+// CompressSnapshot takes uncompressed data as input and compress the data according to Compression Policy
+// and write the compressed data into one end of pipe.
+func CompressSnapshot(data io.ReadCloser, compressionPolicy string) io.ReadCloser {
+	pReader, pWriter := io.Pipe()
+
+	var gWriter io.WriteCloser
+	logger := logrus.New().WithField("actor", "compressor")
+	logger.Infof("start compressing the snapshot using %v Compression Policy", compressionPolicy)
+
+	switch compressionPolicy {
+	case GzipCompressionPolicy:
+		gWriter = gzip.NewWriter(pWriter)
+
+	case LzwCompressionPolicy:
+		gWriter = lzw.NewWriter(pWriter, lzw.LSB, LzwLiteralWidth)
+
+	case ZlibCompressionPolicy:
+		gWriter = zlib.NewWriter(pWriter)
+
+	}
+
+	go func() {
+		var err error
+		defer pWriter.CloseWithError(err)
+		defer gWriter.Close()
```

Comment:
```suggestion
		defer pWriter.CloseWithError(err)
```
Only one of `Close() or `CloseWithError(err)` should be called on the `pWriter`.

Reply:
but `Close() `is called on gWritter not on pWriter and which I think it is necessary to close gWriter as well.

Reply:
Sorry. My bad. You are right.